src/librustc_middle/ty/mod.rs

```diff
@@ -1347,86 +1376,114 @@ impl ToPredicate<'tcx> for PredicateKind<'tcx> {
-        ty::PredicateKind::Trait(self.value.to_poly_trait_predicate(), self.constness)
+        if let Some(pred) = self.value.no_bound_vars() {
+            ty::PredicateAtom::Trait(pred, self.constness).to_predicate(tcx)
+        } else {
+            ty::PredicateKind::ForAll(
+                self.value.map_bound(|pred| {
+                    ty::PredicateAtom::Trait(pred, self.constness).to_predicate(tcx)
+                }),
+            )
             .to_predicate(tcx)
+        }
     }
 }
 
 impl<'tcx> ToPredicate<'tcx> for PolyRegionOutlivesPredicate<'tcx> {
     fn to_predicate(&self, tcx: TyCtxt<'tcx>) -> Predicate<'tcx> {
-        PredicateKind::RegionOutlives(*self).to_predicate(tcx)
+        if let Some(outlives) = self.no_bound_vars() {
+            PredicateAtom::RegionOutlives(outlives).to_predicate(tcx)
+        } else {
+            ty::PredicateKind::ForAll(
+                self.map_bound(|outlives| {
+                    PredicateAtom::RegionOutlives(outlives).to_predicate(tcx)
+                }),
+            )
+            .to_predicate(tcx)
+        }
     }
 }
 
 impl<'tcx> ToPredicate<'tcx> for PolyTypeOutlivesPredicate<'tcx> {
     fn to_predicate(&self, tcx: TyCtxt<'tcx>) -> Predicate<'tcx> {
-        PredicateKind::TypeOutlives(*self).to_predicate(tcx)
+        if let Some(outlives) = self.no_bound_vars() {
+            PredicateAtom::TypeOutlives(outlives).to_predicate(tcx)
+        } else {
+            ty::PredicateKind::ForAll(
+                self.map_bound(|outlives| PredicateAtom::TypeOutlives(outlives).to_predicate(tcx)),
+            )
+            .to_predicate(tcx)
+        }
     }
 }
 
 impl<'tcx> ToPredicate<'tcx> for PolyProjectionPredicate<'tcx> {
     fn to_predicate(&self, tcx: TyCtxt<'tcx>) -> Predicate<'tcx> {
-        PredicateKind::Projection(*self).to_predicate(tcx)
+        if let Some(proj) = self.no_bound_vars() {
+            PredicateAtom::Projection(proj).to_predicate(tcx)
+        } else {
+            ty::PredicateKind::ForAll(
+                self.map_bound(|proj| PredicateAtom::Projection(proj).to_predicate(tcx)),
+            )
+            .to_predicate(tcx)
+        }
     }
 }
 
 impl<'tcx> Predicate<'tcx> {
     pub fn to_opt_poly_trait_ref(self) -> Option<PolyTraitRef<'tcx>> {
-        match self.kind() {
-            &PredicateKind::Trait(ref t, _) => Some(t.to_poly_trait_ref()),
-            PredicateKind::Projection(..)
-            | PredicateKind::Subtype(..)
-            | PredicateKind::RegionOutlives(..)
-            | PredicateKind::WellFormed(..)
-            | PredicateKind::ObjectSafe(..)
-            | PredicateKind::ClosureKind(..)
-            | PredicateKind::TypeOutlives(..)
-            | PredicateKind::ConstEvaluatable(..)
-            | PredicateKind::ConstEquate(..) => None,
+        match self.skip_binders() {
+            PredicateAtom::Trait(t, _) => Some(ty::Binder::bind(t.trait_ref)),
```

Comment:
This I think is only valid because we know there is at most one level of binder, right? Another argument in favor of encoding that into the type system.